torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/Device.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/device_type_analysis.h>
+#include <torch/library.h>
+#include <memory>
+#include <utility>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+using Device = at::Device;
+
+using PropRule = std::function<bool(Node*)>;
+/*
+A Propagation Rule takes the Node, and
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
```

Comment:
I don't know if tracking `changed` here is worth the extra verbosity and difficulty to maintain. I think i commented this somewhere else, but I can't really think of a scenario where this would be useful in the context of dtype/device analysis. We did it in the optimization / graph manipulation passes in order to iteratively optimize.

Reply:
I have removed changed tracking for anywhere that it is not useful. Currently it is needed to know when the types in the for loop construct are consistent. (aka when loop inner start types do not change anymore). 

I have also kept it for the overall pass because that was quite easy.